crates/shared/src/bad_token/token_owner_finder.rs

```diff
@@ -18,99 +19,80 @@ pub trait TokenOwnerFinding: Send + Sync {
-    /// Some nodes struggle with the request and take a long time to respond leading to timeouts.
-    Dynamic,
-}
+    let mut result = Vec::<Arc<dyn TokenOwnerFinding>>::new();
 
-impl UniswapV3Finder {
-    pub async fn new(
-        factory: IUniswapV3Factory,
-        base_tokens: Vec<H160>,
-        current_block: u64,
-        fee_values: FeeValues,
-    ) -> Result<Self> {
-        let fee_values = match fee_values {
-            FeeValues::Static => vec![500, 3000, 10000, 100],
-            // We fetch these once at start up because we don't expect them to change often.
-            // Alternatively could use a time based cache.
-            FeeValues::Dynamic => Self::fee_values(&factory, current_block).await?,
-        };
-        tracing::debug!(?fee_values);
-        Ok(Self {
-            factory,
-            base_tokens,
-            fee_values,
-        })
+    if finders.contains(&TokenOwnerFindingStrategy::Liquidity) {
+        result.extend(
+            pair_providers
+                .iter()
+                .map(|provider| -> Arc<dyn TokenOwnerFinding> {
+                    Arc::new(UniswapLikePairProviderFinder {
+                        inner: provider.clone(),
+                        base_tokens: base_tokens.tokens().iter().copied().collect(),
+                    })
+                }),
+        );
+        if let Some(contract) = vault {
+            result.push(Arc::new(BalancerVaultFinder(contract.clone())));
+        }
+        if let Some(contract) = uniswapv3_factory {
+            result.push(Arc::new(
+                UniswapV3Finder::new(
+                    contract.clone(),
+                    base_tokens.tokens().iter().copied().collect(),
+                    current_block,
+                    uniswapv3_fee_values,
+                )
+                .await?,
+            ));
+        }
     }
 
-    // Possible fee values as given by
-    // https://github.com/Uniswap/v3-core/blob/9161f9ae4aaa109f7efdff84f1df8d4bc8bfd042/contracts/UniswapV3Factory.sol#L26
-    async fn fee_values(factory: &IUniswapV3Factory, current_block: u64) -> Result<Vec<u32>> {
-        // We expect there to be few of these kind of events (currently there are 4) so fetching all
-        // of them is fine. Alternatively we could index these events in the database.
-        let events = factory
-            .events()
-            .fee_amount_enabled()
-            .from_block(BlockNumber::Earliest)
-            .to_block(BlockNumber::Number(
-                current_block.saturating_sub(MAX_REORG_BLOCK_COUNT).into(),
-            ))
-            .query()
-            .await?;
-        let fee_values = events.into_iter().map(|event| event.data.fee).collect();
-        Ok(fee_values)
+    if finders.contains(&TokenOwnerFindingStrategy::Liquidity) {
+        if let Ok(finder) = BlockscoutTokenOwnerFinder::try_with_network(client.clone(), chain_id) {
+            result.push(Arc::new(finder));
+        }
```

Comment:
I know this was already missing before the refactor but should we maybe have a log here when initialization fails?

Reply:
Actually, the `init` returns a `Result` already - it should just fail to initialize the finders.

Reply:
Fixed.